### PR TITLE
[NXCM-3121] Filter out repositories for automatic selection based on user

### PR DESCRIPTION
**Pull request by adreghiciu**
[NXCM-3121] Filter out repositories for automatic selection based on user ip and user agent

To do so we make maven resolve an artifact from a "special" repository, handled by a rest resource. The response is a property file containing info about whom made the call. If the artifact could be resolved we then use that info to filter out the repositories for automatic selection.

Signed-off-by: Alin Dreghiciu adreghiciu@gmail.com

### CODE REVIEW

File: nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java [hunk 1]
```diff
@@ -234,6 +287,117 @@ public abstract class AbstractStagingMojo
         }
     }
 
+    protected Identity whoAmI()
```
Comment (velo) [inline nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java:299]: Missing jdoc

**Comment by cstamas** [comment]
Looks unbelievably involved just to perform so simple thing... also looks very fragile to me.

But as we went over this multiple times... if there is really no proper way than "pinging" remote to get the needed info... :(

**Comment by velo** [comment]
A topic for a future discussion, this maven plugin may should be part of nexus pro, since the fix is partially here and partially there...
https://github.com/sonatype/nexus-enterprise/pull/24/files

File: nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java (hunk 1 above)
Comment (adreghiciu) [inline nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java:299]: Added in c75d165b

File: nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java [hunk 2]
```diff
@@ -234,6 +287,126 @@ public abstract class AbstractStagingMojo
...
+                "local", createTempFile( "whoami-", "", null ).toURI().toASCIIString(), defaultRepositoryLayout
+            );
+
+            // looks like Maven 2 does not give any chance of artifact provided username/password so try a fallback
+            final AuthenticationInfo backupAuthInfo = wagonManager.getAuthenticationInfo( whoAmIRepository.getId() );
+            wagonManager.addAuthenticationInfo( whoAmIRepository.getId(), getUsername(), getPassword(), null, null );
+
+            try
+            {
+                artifactResolver.resolve( whoAmIArtifact, asList( whoAmIRepository ), fakeLocal );
```
Comment (velo) [inline nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java:328]: I assume
artifactResolver.resolve( whoAmIArtifact, asList( whoAmIRepository ), fakeLocal );
is used so nexus can tell you the user agent, is that it?

**Comment by adreghiciu** [comment]
@cstamas Do you see an alternative approach? 
For IP we must for sure make a nexus call, but that could be done without involving Maven.
For User-Agent is in my view less error prone then trying to dig it out from Maven.

And even if we run into problems, such as nexus not responding, maven not being able to connect, ..., we then just fall back to current approach = we will not filter out the repositories based on user ip/agent.

**Comment by velo** [comment]
Far enough
+1

File: nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java (hunk 2 above)
Comment (adreghiciu) [inline nexus-maven-plugins/nexus-maven-plugin/src/main/java/org/sonatype/nexus/plugin/AbstractStagingMojo.java:328]: user agent and user ip

**Pull request closed without merging**